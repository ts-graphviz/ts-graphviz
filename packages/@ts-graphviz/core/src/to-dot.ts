import { DotObjectModel } from '@ts-graphviz/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '@ts-graphviz/ast';

/**
 * This interface provides options for converting a model to DOT.
 * @group Convert Model to DOT
 */
export interface ToDotOptions {
  /**
   * Options for converting the model to DOT.
   */
  convert?: ConvertFromModelOptions;
  /**
   * Options for printing DOT.
   */
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Subgraph}, {@link Node}, and {@link Edge}
 * @param options Optional options for the conversion.
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
