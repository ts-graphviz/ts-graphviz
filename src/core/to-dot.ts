import { DotObjectModel } from '#lib/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '#lib/ast';

/**
 * @group Convert Model to DOT
 * @alpha
 */
export interface ToDotOptions {
  convert?: ConvertFromModelOptions;
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Node}, and {@link Edge}
 * @param options
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
