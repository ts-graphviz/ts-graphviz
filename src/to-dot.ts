import { DotObjectModel } from './common/index.js';
import { stringify, fromModel, PrintOptions } from './ast/index.js';
import { ConvertOptions } from './ast/index.js';

export interface ToDotOptions {
  convert?: ConvertOptions;
  print?: PrintOptions;
}

export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
