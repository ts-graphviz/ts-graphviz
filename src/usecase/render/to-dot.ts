import { Renderer, Dot } from './renderer.js';

export function toDot(object: Dot): string {
  const renderer = new Renderer();
  return renderer.render(object);
}
