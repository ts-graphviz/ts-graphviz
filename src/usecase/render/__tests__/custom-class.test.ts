import { attribute } from '../../../attribute/index.js';
import { toDot } from '../to-dot.js';
import { Graph, Node, Edge, EdgeTargetTuple } from '../../../model/index.js';

describe('class base API', () => {
  test('digraph', () => {
    const digraph = new Graph(true);
    const node1 = new Node('node1', {
      [attribute.label]: 'Node 1 label',
    });
    const node2 = new Node('node2', {
      [attribute.label]: 'Node 2 label',
    });
    const edge = new Edge([node1, node2], {
      [attribute.label]: 'Edge label',
    });
    digraph.addNode(node1);
    digraph.addNode(node2);
    digraph.addEdge(edge);
    expect(toDot(digraph)).toMatchSnapshot();
  });

  test('graph', () => {
    const graph = new Graph(false);
    const node1 = new Node('node1', {
      [attribute.label]: 'Node 1 label',
    });
    const node2 = new Node('node2', {
      [attribute.label]: 'Node 2 label',
    });
    const edge = new Edge([node1, node2], {
      [attribute.label]: 'Edge label',
    });
    graph.addNode(node1);
    graph.addNode(node2);
    graph.addEdge(edge);
    expect(toDot(graph)).toMatchSnapshot();
  });

  describe('custom class', () => {
    class MyCustomNode extends Node {
      constructor(id: number) {
        super(`node${id}`, {
          [attribute.label]: `This is Custom Node ${id}`,
        });
      }
    }

    class MyCustomEdge extends Edge {
      constructor(targets: EdgeTargetTuple) {
        super(targets, {
          [attribute.label]: `This is Custom Edge`,
        });
      }
    }

    class MyCustomDigraph extends Graph {
      constructor() {
        super(true, 'G', {
          [attribute.label]: 'This is Custom Digraph',
        });
      }
    }

    class MyCustomGraph extends Graph {
      constructor() {
        super(false, 'G', {
          [attribute.label]: 'This is Custom Graph',
        });
      }
    }

    test('render custom node', () => {
      const digraph = new Graph(true);
      const node = new MyCustomNode(1);
      digraph.addNode(node);
      expect(toDot(digraph)).toMatchSnapshot();
    });

    test('render custom digraph', () => {
      const digraph = new MyCustomDigraph();
      expect(toDot(digraph)).toMatchSnapshot();
    });

    test('render custom graph', () => {
      const graph = new MyCustomGraph();
      expect(toDot(graph)).toMatchSnapshot();
    });

    test('render custom edge', () => {
      const digraph = new Graph(true);
      const node1 = new Node('node1', {
        [attribute.label]: 'Node 1 label',
      });
      const node2 = new Node('node2', {
        [attribute.label]: 'Node 2 label',
      });
      const edge = new MyCustomEdge([node1, node2]);
      digraph.addNode(node1);
      digraph.addNode(node2);
      digraph.addEdge(edge);
      expect(toDot(digraph)).toMatchSnapshot();
    });

    test('render custom objects', () => {
      const digraph = new MyCustomDigraph();
      const node1 = new MyCustomNode(1);
      const node2 = new MyCustomNode(2);
      const edge = new MyCustomEdge([node1, node2]);
      digraph.addNode(node1);
      digraph.addNode(node2);
      digraph.addEdge(edge);
      expect(toDot(digraph)).toMatchSnapshot();
    });
  });
});
