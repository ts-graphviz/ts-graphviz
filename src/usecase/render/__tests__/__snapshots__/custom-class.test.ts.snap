// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class base API custom class render custom digraph 1`] = `
"digraph \\"G\\" {
  label = \\"This is Custom Digraph\\";
}"
`;

exports[`class base API custom class render custom edge 1`] = `
"digraph {
  \\"node1\\" [
    label = \\"Node 1 label\\",
  ];
  \\"node2\\" [
    label = \\"Node 2 label\\",
  ];
  \\"node1\\" -> \\"node2\\" [
    label = \\"This is Custom Edge\\",
  ];
}"
`;

exports[`class base API custom class render custom graph 1`] = `
"graph \\"G\\" {
  label = \\"This is Custom Graph\\";
}"
`;

exports[`class base API custom class render custom node 1`] = `
"digraph {
  \\"node1\\" [
    label = \\"This is Custom Node 1\\",
  ];
}"
`;

exports[`class base API custom class render custom objects 1`] = `
"digraph \\"G\\" {
  label = \\"This is Custom Digraph\\";
  \\"node1\\" [
    label = \\"This is Custom Node 1\\",
  ];
  \\"node2\\" [
    label = \\"This is Custom Node 2\\",
  ];
  \\"node1\\" -> \\"node2\\" [
    label = \\"This is Custom Edge\\",
  ];
}"
`;

exports[`class base API digraph 1`] = `
"digraph {
  \\"node1\\" [
    label = \\"Node 1 label\\",
  ];
  \\"node2\\" [
    label = \\"Node 2 label\\",
  ];
  \\"node1\\" -> \\"node2\\" [
    label = \\"Edge label\\",
  ];
}"
`;

exports[`class base API graph 1`] = `
"graph {
  \\"node1\\" [
    label = \\"Node 1 label\\",
  ];
  \\"node2\\" [
    label = \\"Node 2 label\\",
  ];
  \\"node1\\" -- \\"node2\\" [
    label = \\"Edge label\\",
  ];
}"
`;
